### ISSUE TITLE
Feature: Code Mode for MCP - single endpoint with TypeScript execution

### ISSUE THREAD

**Issue by jacob-petterle** [SchoolAI/peer-plan#46]
## Summary

Consolidate peer-plan's 10+ separate MCP tool endpoints into a single **code execution endpoint** that allows agents to write TypeScript code calling the peer-plan API directly. This dramatically reduces token overhead and improves efficiency based on Cloudflare and Anthropic's research.

## Problem

**Current architecture (tool-based):**
```
create_plan() ← separate tool call
add_artifact() ← separate tool call
post_progress_update() ← separate tool call
read_plan() ← separate tool call
link_pr() ← separate tool call
...etc
```

**Issues:**
- Each tool call adds LLM overhead
- Context window bloated with tool definitions
- Intermediate results fed back through LLM
- Lots of token consumption for complex workflows
- Harder for agent to compose multi-step operations
- Agent must remember all tool names and signatures

## Proposed Solution

Expose peer-plan as a **single TypeScript API endpoint** where agents write code:

```typescript
// One tool: execute_code()
execute_code({
  code: `
    // Agent writes TypeScript that calls the API
    const plan = await peerprogramming.createPlan({
      title: 'Authentication system',
      content: '...',
    });
    
    // Can use loops, conditionals, error handling
    const artifacts = [...];
    for (const artifact of artifacts) {
      await peerprogramming.addArtifact(plan.id, artifact);
    }
    
    // Link PR
    await peerprogramming.linkPR(plan.id, 42);
    
    // Get feedback
    const feedback = await peerprogramming.readPlan(plan.id, { 
      includeAnnotations: true 
    });
    
    return feedback;
  `
})
```

### Why This Works Better

Per [Cloudflare's Code Mode research](https://blog.cloudflare.com/code-mode/):
- **81% token reduction** for complex batch operations
- **LLMs are better at writing code** than calling tools directly (more training data)
- **Eliminates intermediate processing** - code runs in sandbox, results don't feed back through neural network
- **Single endpoint** - Reduces context bloat from tool definitions

Per [Anthropic's Code Execution with MCP](https://www.anthropic.com/engineering/code-execution-with-mcp):
- **98.7% cost savings** in some scenarios (150k → 2k tokens\!)
- **Progressive disclosure** - Agent discovers and loads only needed functions
- **Preserves performance** - Loops and conditionals execute in code, not in LLM loop

### Current (Tool-Based)

```
Agent 
  ├─ Call: create_plan() → LLM processes result
  ├─ Call: add_artifact() → LLM processes result
  ├─ Call: post_progress_update() → LLM processes result
  └─ ...etc (each call goes through neural network)

Token usage: ~150k for complex workflows
Context window: Bloated with 10+ tool definitions
```

### Proposed (Code Mode)

```
Agent 
  └─ Call: execute_code(agentWritesTypeScript)
      ↓
    Sandboxed Worker
      ├─ Runs TypeScript
      ├─ Calls peerprogramming.* functions
      ├─ Executes loops/conditionals/logic
      └─ Returns final result to Agent

Token usage: ~2k for same workflow
Context window: Only ONE tool definition
```

### Phase 1: Expose TypeScript API

Create `peer-plan-sdk` module that exposes functions:

```typescript
// peer-plan-sdk/src/index.ts

export const peerprogramming = {
  // Existing tool functions become SDK methods
  createPlan: async (args: CreatePlanInput) => { ... },
  readPlan: async (planId: string, opts?: ReadPlanOpts) => { ... },
  updatePlan: async (planId: string, updates: PlanUpdates) => { ... },
  addArtifact: async (planId: string, artifact: Artifact) => { ... },
  linkPR: async (planId: string, prNumber: number) => { ... },
  postProgressUpdate: async (planId: string, update: ProgressUpdate) => { ... },
  // ...etc
};

export default peerprogramming;
```

### Phase 2: Single MCP Tool Endpoint

Add ONE tool to MCP server:

```typescript
server.tool(
  'execute_code',
  {
    code: z.string().describe("TypeScript code that uses peerprogramming API"),
    timeout: z.number().optional().describe("Max execution time in ms"),
  },
  async (args) => {
    // Create isolated sandbox
    const sandbox = new Sandbox({
      imports: { peerprogramming },
      timeout: args.timeout || 30000,
    });
    
    // Execute agent-written code
    const result = await sandbox.execute(args.code);
    
    return { 
      success: true, 
      result,
      tokensUsed: calculateTokens(result),
    };
  }
);
```

### Phase 3: Update MCP Tools (Parallel)

Keep old tools for backward compatibility, but mark as deprecated:
- `create_plan` → Use `execute_code(peerprogramming.createPlan(...))` instead
- `add_artifact` → Use within `execute_code`
- ...etc

Provide migration guide for agents.

## Sandbox Implementation

Use Cloudflare Workers or Node.js VM2 for sandboxing:

```typescript
import { VM } from 'vm2';

class Sandbox {
  constructor(options: { imports: object, timeout: number }) {
    this.vm = new VM({
      timeout: options.timeout,
      sandbox: options.imports,
    });
  }
  
  async execute(code: string) {
    return this.vm.run(`
      (async () => {
        ${code}
      })()`
    );
  }
}
```

**Important:** Sandbox must have:
- ✅ Access to peer-plan functions
- ✅ Ability to write TypeScript (loops, conditionals, error handling)
- ❌ No access to general internet
- ❌ No access to file system (except peer-plan context)
- ❌ No access to other agents' data

### Workflow 1: Create Plan with Artifacts (Current: 4 tool calls → New: 1 call)

**Current (Tool-Based):**
```
Agent: Call create_plan(...)
       ↓ Get planId
Agent: Call add_artifact(planId, screenshot)
       ↓ Get artifactUrl
Agent: Call add_artifact(planId, test-results)
       ↓ Get artifactUrl
Agent: Call link_pr(planId, 42)
       ↓ Get status
Total: 4 tool calls, high token usage
```

**Proposed (Code Mode):**
```
Agent: Call execute_code(`
  const plan = await peerprogramming.createPlan({...});
  await peerprogramming.addArtifact(plan.id, screenshot);
  await peerprogramming.addArtifact(plan.id, testResults);
  await peerprogramming.linkPR(plan.id, 42);
  return plan;
`)
Total: 1 tool call, minimal overhead
```

### Workflow 2: Batch Operations with Conditions

Code mode excels at complex logic:

```typescript
execute_code(`
  const plans = [];
  
  // Loop without feeding back to LLM
  for (const requirement of requirements) {
    const plan = await peerprogramming.createPlan({
      title: requirement.name,
      content: requirement.spec,
    });
    
    // Conditional logic in code
    if (requirement.priority === 'high') {
      await peerprogramming.updatePlan(plan.id, { 
        status: 'pending_review' 
      });
    }
    
    // Error handling
    try {
      await peerprogramming.addArtifact(plan.id, requirement.docs);
    } catch (e) {
      console.log(`Failed to add artifact: ${e.message}`);
    }
    
    plans.push(plan);
  }
  
  return plans;
`)
```

### Workflow 3: Conditional Logic Based on Feedback

```typescript
execute_code(`
  const plan = await peerprogramming.readPlan(planId, { 
    includeAnnotations: true 
  });
  
  // Check feedback
  const hasBlockers = plan.activity.some(u => u.type === 'blocker');
  
  if (hasBlockers) {
    // Update status to blocked
    await peerprogramming.updatePlan(planId, { 
      status: 'blocked' 
    });
    return 'Plan blocked, awaiting reviewer input';
  } else {
    // Continue with implementation
    await peerprogramming.postProgressUpdate(planId, {
      type: 'checkpoint',
      title: 'Ready to implement',
    });
    return 'Proceeding with implementation';
  }
`)
```

### Week 1: Foundation
- [ ] Create peer-plan-sdk with all functions
- [ ] Implement sandbox (VM2 or Workers)
- [ ] Add execute_code tool to MCP server
- [ ] Write tests for sandbox execution

### Week 2: Integration
- [ ] Update CLAUDE.md with code mode examples
- [ ] Add error handling for sandbox timeouts
- [ ] Document migration from tool-based to code mode
- [ ] Update MCP server docs

### Week 3: Deprecation
- [ ] Mark old tools as deprecated (but keep working)
- [ ] Add warnings when using old tools
- [ ] Provide migration script if needed
- [ ] Update all examples/docs

### Future
- [ ] Consider keeping tool-based for simple operations (backward compat)
- [ ] Add code mode to other SDKs (Python, Go)

### Sandbox Isolation
- ✅ Code runs in isolated VM/Worker
- ✅ Can't access file system
- ✅ Can't make arbitrary HTTP requests
- ✅ Can't access other agents' plans (permissions enforced by peer-plan API)

### API Keys
- Sandbox gets peerprogramming API access only
- No other credentials exposed
- Each sandbox is disposable (destroyed after execution)

### Rate Limiting
- Limit code execution time (30s timeout)
- Limit sandbox memory (256MB)
- Limit request rate per agent

## Acceptance Criteria

- [ ] execute_code tool accepts TypeScript code
- [ ] Code sandbox isolates execution safely
- [ ] Can call all peer-plan SDK functions from code
- [ ] Complex workflows run with single tool call
- [ ] Code execution produces expected results
- [ ] Backward compatibility with old tools (deprecated but working)
- [ ] Error handling works (timeouts, runtime errors)
- [ ] Token usage is measurably lower than tool-based approach
- [ ] Documentation and examples updated

## Performance Targets

| Metric | Old (Tool-Based) | New (Code Mode) | Target |
|--------|-----------------|-----------------|--------|
| Token usage | 150k | 2k | 98% reduction |
| Latency | 5s (5 tool calls × 1s) | 3s (1 code call) | 40% faster |
| Context bloat | 10+ tool definitions | 1 tool definition | Much cleaner |

## References

- [Cloudflare: Code Mode - the better way to use MCP](https://blog.cloudflare.com/code-mode/)
- [Anthropic: Code Execution with MCP](https://www.anthropic.com/engineering/code-execution-with-mcp)
- [VM2 Sandbox Library](https://github.com/patriksimek/vm2)
- [Cloudflare Agents: Code Mode Implementation](https://developers.cloudflare.com/agents/model-context-protocol/)

## Why This is P1

- **Massive efficiency gain** - 81-98% token reduction for complex workflows
- **Better agent experience** - Agents write code naturally instead of juggling tools
- **Scales better** - Single endpoint vs 10+ tools
- **Future-proof** - Aligns with Cloudflare and Anthropic's recommended patterns
- **Unblocks advanced workflows** - Loops, conditionals, error handling in code
- **Reduces tool calling burden** - One tool instead of memorizing many

## Related Issues

- #17 - Claude Code integration (will benefit greatly from code mode)
- #39 - Activity feed (can be populated with code-based batch operations)
- #41 - Teleport context (entire operation can be one code call)

---

*Generated from discussion on 2026-01-08, corrected understanding of Code Mode*
